cloud/vmware/vmware_add_hdd.py

```diff
@@ -0,0 +1,211 @@
+        new_disk_kb = int(disk_size) * 1024 * 1024
+        disk_spec = vim.vm.device.VirtualDeviceSpec()
+        disk_spec.fileOperation = "create"
+        disk_spec.operation = vim.vm.device.VirtualDeviceSpec.Operation.add
+        disk_spec.device = vim.vm.device.VirtualDisk()
+        disk_spec.device.backing = vim.vm.device.VirtualDisk.FlatVer2BackingInfo()
+        if thinDisk is True:
+            disk_spec.device.backing.thinProvisioned = True
+        disk_spec.device.backing.diskMode = 'persistent'
+        disk_spec.device.unitNumber = unit_number
+        disk_spec.device.capacityInKB = new_disk_kb
+        disk_spec.device.controllerKey = controller.key
+        dev_changes.append(disk_spec)
+        spec.deviceChange = dev_changes
+        vm.ReconfigVM_Task(spec=spec)
+
+        return unit_number, controller.key
+
+
+def objwalk(obj, path_elements):
+    if hasattr(obj, 'parent'):
+        new_obj = getattr(obj, 'parent')
+        if new_obj:
+            if new_obj.name != 'vm':
+                # 'vm' is an invisible folder that exists at the datacenter root so we ignore it
+                path_elements.append(new_obj.name)
+
+            objwalk(new_obj, path_elements)
+
+    return path_elements
+
+
+def get_vm_object(module, conn, path):
+
+    all_vms = get_all_objs(conn, [vim.VirtualMachine])
+    matching_vms = []
+    path_list = filter(None, path.split('/'))
+    name = path_list.pop()
+
+    for vm_obj, label in all_vms.iteritems():
+        if label == name:
+            matching_vms.append(vm_obj)
+
+    try:
+        if len(matching_vms) > 1:
+
+            for vm_obj in matching_vms:
+                elements = []
+                if set(path_list).issubset(set(objwalk(vm_obj, elements))):
+                    return vm_obj
+
+        else:
+            return matching_vms[0]
+
+    except TypeError:
+        module.fail_json(msg='Could not find VM at %s' % path)
+
+
+def main():
+
+    argument_spec = vmware_argument_spec()
+
+    argument_spec.update(
+        dict(
+            thinDisk=dict(required=False, default=True, type='bool'),
+            vm=dict(required=True, type='str'),
+            diskSize=dict(required=True, type='int'),
+        )
+    )
+
+    module = AnsibleModule(argument_spec=argument_spec)
+
+    if HAS_PYVMOMI == False:
+        module.fail_json(msg='pyvmomi is required for this module')
+
+    hostname = module.params['hostname']
+    username = module.params['username']
+    password = module.params['password']
+    thinDisk = module.params['thinDisk']
+    vm = module.params['vm']
+    diskSize = module.params['diskSize']
+    validate_certs = module.params['validate_certs']
+
+    conn = connect_to_api(module)
+    vmToWorkOn = get_vm_object(module, conn, vm)
+
+    try:
+        unitNumber, controlerKey = add_disk(module, vmToWorkOn, diskSize, thinDisk)
+        module.exit_json(changed=True, vm=vm, diskUnitNumber=unitNumber, controlerKey=controlerKey, isThinDisk=thinDisk)
```

Comment:
While I may sound pedantic, shouldn't it be controllerKey ?(as I think it might confuse people, but I am also not a native english speaker)